Ambiguous conversion from nlohmann::basic_json<> to custom class.

Compile time error occurs if I try to convert instance of `nlohmann::basic_json<>` to instance of class which has converting constructor with template specialization for `nlohmann::basic_json<>`. This regression is present in v3.10.4, v3.10.5 and develop (d8a6329).

Code example, which expose this issue - `nlohmann-json-issue.cpp`:
```c++
#include <iostream>
#include <stdexcept>
#include <nlohmann/json.hpp>

class Point final
{
public:
    Point(int x = 0, int y = 0, int z = 0);

    template <class T>
    Point(const T&);

    void Print() const;

    int x;
    int y;
    int z;
};

template <>
Point::Point(const nlohmann::json& json):
   Point(json.value("x", 0), json.value("y", 0), json.value("z", 0)) 
{
}

int main(void)
{
    const nlohmann::json pointJson = {
        {"x",111},
        {"y",222},
        {"z",333}
    };

    Point p = pointJson; // <- ambiguous conversion !!!
    p.Print();

    return 0;
}

Point::Point(int x, int y, int z): x(x), y(y), z(z)
{
}

template <class T>
Point::Point(const T&)
{
    throw std::logic_error("Not implemented");
}

void Point::Print() const
{
   std::cout << "(" << x << "," << y << "," << z << ")" << std::endl;
}
```

Compile error:
```
[pavel@well cpp-test]$ g++ nlohmann-json-issue.cpp -o nlohmann-json-issue
nlohmann-json-issue.cpp: In function ‘int main()’:
nlohmann-json-issue.cpp:33:15: error: conversion from ‘const json’ {aka ‘const nlohmann::basic_json<>’} to ‘Point’ is ambiguous
   33 |     Point p = pointJson; // <- ambiguous conversion !!!
      |               ^~~~~~~~~
In file included from nlohmann-json-issue.cpp:2:
/usr/include/nlohmann/json.hpp:1885:55: note: candidate: ‘nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType>::operator ValueType() const [with ValueType = Point; typename std::enable_if<nlohmann::detail::conjunction<nlohmann::detail::negation<std::is_pointer<_Ptr> >, nlohmann::detail::negation<std::is_same<ValueType, nlohmann::detail::json_ref<nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType> > > >, nlohmann::detail::negation<std::is_same<ValueType, typename StringType::value_type> >, nlohmann::detail::negation<nlohmann::detail::is_basic_json<BasicJsonType> >, nlohmann::detail::negation<std::is_same<ValueType, std::initializer_list<typename StringType::value_type> > >, nlohmann::detail::negation<std::is_same<ValueType, std::basic_string_view<char, std::char_traits<char> > > >, nlohmann::detail::is_detected_lazy<nlohmann::detail::get_template_function, const nlohmann::basic_json<ObjectType, ArrayType, StringType, BooleanType, NumberIntegerType, NumberUnsignedType, NumberFloatType, AllocatorType, JSONSerializer, BinaryType>&, ValueType> >::value, int>::type <anonymous> = 0; ObjectType = std::map; ArrayType = std::vector; StringType = std::__cxx11::basic_string<char>; BooleanType = bool; NumberIntegerType = long int; NumberUnsignedType = long unsigned int; NumberFloatType = double; AllocatorType = std::allocator; JSONSerializer = nlohmann::adl_serializer; BinaryType = std::vector<unsigned char>]’
 1885 |                                         JSON_EXPLICIT operator ValueType() const
      |                                                       ^~~~~~~~
nlohmann-json-issue.cpp:20:1: note: candidate: ‘Point::Point(const T&) [with T = nlohmann::basic_json<>]’
   20 | Point::Point(const nlohmann::json& json):
      | ^~~~~
```

Workarounds:
 - compile with `-DJSON_USE_IMPLICIT_CONVERSIONS=0`;
 - make variable  `pointJson` mutable;

Compiler: g++ (GCC) 11.1.0
OS: Arch Linux

--- a/.clang-tidy
+++ b/.clang-tidy
@@ -43,7 +43,7 @@ Checks: '*,
          -readability-identifier-length,
          -readability-magic-numbers,
          -readability-redundant-access-specifiers,
-         -readability-simplify-boolean-expr
+         -readability-simplify-boolean-expr,
          -readability-uppercase-literal-suffix'
 
 CheckOptions:

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -270,6 +270,56 @@ inline void from_json(const json& j, for_3312& obj)
 }
 #endif
 
+/////////////////////////////////////////////////////////////////////
+// for #3204
+/////////////////////////////////////////////////////////////////////
+
+struct for_3204_foo
+{
+    for_3204_foo() = default;
+    for_3204_foo(std::string /*unused*/) {}
+};
+
+struct for_3204_bar
+{
+    enum constructed_from_t
+    {
+        constructed_from_none = 0,
+        constructed_from_foo = 1,
+        constructed_from_json = 2
+    };
+
+    for_3204_bar(std::function<void(for_3204_foo)> /*unused*/) // NOLINT(performance-unnecessary-value-param)
+        : constructed_from(constructed_from_foo) {}
+    for_3204_bar(std::function<void(json)> /*unused*/) // NOLINT(performance-unnecessary-value-param)
+        : constructed_from(constructed_from_json) {}
+
+    constructed_from_t constructed_from = constructed_from_none;
+};
+
+/////////////////////////////////////////////////////////////////////
+// for #3333
+/////////////////////////////////////////////////////////////////////
+
+struct for_3333 final
+{
+    for_3333(int x_ = 0, int y_ = 0) : x(x_), y(y_) {}
+
+    template <class T>
+    for_3333(const T& /*unused*/)
+    {
+        CHECK(false);
+    }
+
+    int x = 0;
+    int y = 0;
+};
+
+template <>
+inline for_3333::for_3333(const json& json)
+    : for_3333(json.value("x", 0), json.value("y", 0))
+{}
+
 TEST_CASE("regression tests 2")
 {
     SECTION("issue #1001 - Fix memory leak during parser callback")
@@ -862,6 +912,28 @@ TEST_CASE("regression tests 2")
         CHECK(a2.type() == typeid(j));
     }
 #endif
+
+    SECTION("issue #3204 - ambiguous regression")
+    {
+        for_3204_bar bar_from_foo([](for_3204_foo) {});
+        for_3204_bar bar_from_json([](json) {});
+
+        CHECK(bar_from_foo.constructed_from == for_3204_bar::constructed_from_foo);
+        CHECK(bar_from_json.constructed_from == for_3204_bar::constructed_from_json);
+    }
+
+    SECTION("issue #3333 - Ambiguous conversion from nlohmann::basic_json<> to custom class")
+    {
+        const json j
+        {
+            {"x", 1},
+            {"y", 2}
+        };
+        for_3333 p = j;
+
+        CHECK(p.x == 1);
+        CHECK(p.y == 2);
+    }
 }
 
 DOCTEST_CLANG_SUPPRESS_WARNING_POP

--- a/tests/src/unit-regression2.cpp
+++ b/tests/src/unit-regression2.cpp
@@ -277,7 +277,7 @@ inline void from_json(const json& j, for_3312& obj)
 struct for_3204_foo
 {
     for_3204_foo() = default;
-    for_3204_foo(std::string /*unused*/) {}
+    explicit for_3204_foo(std::string /*unused*/) {} // NOLINT(performance-unnecessary-value-param)
 };
 
 struct for_3204_bar
@@ -289,9 +289,9 @@ struct for_3204_bar
         constructed_from_json = 2
     };
 
-    for_3204_bar(std::function<void(for_3204_foo)> /*unused*/) // NOLINT(performance-unnecessary-value-param)
+    explicit for_3204_bar(std::function<void(for_3204_foo)> /*unused*/) noexcept // NOLINT(performance-unnecessary-value-param)
         : constructed_from(constructed_from_foo) {}
-    for_3204_bar(std::function<void(json)> /*unused*/) // NOLINT(performance-unnecessary-value-param)
+    explicit for_3204_bar(std::function<void(json)> /*unused*/) noexcept // NOLINT(performance-unnecessary-value-param)
         : constructed_from(constructed_from_json) {}
 
     constructed_from_t constructed_from = constructed_from_none;
@@ -316,8 +316,8 @@ struct for_3333 final
 };
 
 template <>
-inline for_3333::for_3333(const json& json)
-    : for_3333(json.value("x", 0), json.value("y", 0))
+inline for_3333::for_3333(const json& j)
+    : for_3333(j.value("x", 0), j.value("y", 0))
 {}
 
 TEST_CASE("regression tests 2")
@@ -915,8 +915,8 @@ TEST_CASE("regression tests 2")
 
     SECTION("issue #3204 - ambiguous regression")
     {
-        for_3204_bar bar_from_foo([](for_3204_foo) {});
-        for_3204_bar bar_from_json([](json) {});
+        for_3204_bar bar_from_foo([](for_3204_foo) noexcept {}); // NOLINT(performance-unnecessary-value-param)
+        for_3204_bar bar_from_json([](json) noexcept {}); // NOLINT(performance-unnecessary-value-param)
 
         CHECK(bar_from_foo.constructed_from == for_3204_bar::constructed_from_foo);
         CHECK(bar_from_json.constructed_from == for_3204_bar::constructed_from_json);

